### PR TITLE
[WIP] Import ONNX models

### PR DESCRIPTION
**Pull request by yg2288**
#29 Support importing models from ONNX

### CODE REVIEW

File: dlpy/tests/test_model.py [hunk 1]
```diff
@@ -617,6 +618,18 @@ def test_model23(self):
 
         model1.deploy(self.data_dir, output_format='onnx')
 
+    def test_model24(self):
```
Comment (ethem-kinginthenorth) [inline dlpy/tests/test_model.py]: I was wondering if you can add another onnx example for tests. I can't evaluate the coverage but two would cover more than one.

**Comment by yg2288** [comment]
Working on adding more tests and comments, thanks.

File: dlpy/tests/test_model.py (hunk 1 above)
Comment (yg2288) [reply dlpy/tests/test_model.py]: Added additional test cases

**Comment by yg2288** [comment]
Tests are passing

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+
+def onnx_initializer_to_tensors(initializer):
+    ''' Convert ONNX graph initializer to tensors '''
+
+    return [(init.name,
+             numpy_helper.to_array(init))
+            for init in initializer]
+
+
+def onnx_filter_sas_layers(graph):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+    ''' Filter nodes that correspond to SAS Deep learning layer types '''
+
+    sas_layers = [node for node in graph.node
+                  if is_compute_layer(graph, node)]
+
+    return sas_layers
+
+
+def onnx_input_layer(value_info):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 4]
```diff
@@ -0,0 +1,855 @@
...
+        return onnx_extract_residual(graph, node, layers)
+    elif node.op_type == 'GlobalAveragePool':
+        return onnx_extract_globalpool(graph, node, layers)
+    else:
+        raise OnnxParseError('Unsupported ONNX op: '
+                             + str(node.name) + ', '
+                             + str(node.op_type))
+
+
+def find_input_layer_name(graph):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+    uninitialized = [value_info.name for value_info in graph.input
+                     if value_info.name not in initialized]
+    if not uninitialized:
+        raise OnnxParseError('Unable to determine input layer.')
+    if len(uninitialized) > 1:
+        raise OnnxParseError('Unable to determine input layer.')
+    return uninitialized[0]
+
+
+def get_dlpy_layer(layers, name):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+    ''' Get a layer by name from list of layers ''' 
+    for layer in layers:
+        if layer.name == name:
+            return layer
+    return None
+
+
+def onnx_extract_conv(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+                  stride_vertical=stride_vertical,
+                  padding=padding,
+                  padding_width=padding_width,
+                  padding_height=padding_height,
+                  act=act,
+                  include_bias=include_bias,
+                  src_layers=src) 
+
+    
+def onnx_extract_pool(graph, node, layers, pool='MAX'):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+                   stride_horizontal=stride_horizontal,
+                   stride_vertical=stride_vertical,
+                   padding=padding,
+                   padding_width=padding_width,
+                   padding_height=padding_height,
+                   pool=pool,
+                   src_layers=src)
+
+
+def onnx_extract_globalpool(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+    return Pooling(width=width,
+                   height=height,
+                   stride=1,
+                   name=node.name,
+                   padding=0,
+                   pool='AVERAGE',
+                   src_layers=src)
+
+
+def onnx_extract_batchnormalization(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+        src_names = [p.name for p in previous]
+
+    src = [get_dlpy_layer(layers, i) for i in src_names]
+
+    return BN(name=node.name,
+              act='identity',
+              src_layers=src)
+
+
+def onnx_extract_concat(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 11]
```diff
@@ -0,0 +1,855 @@
...
+        src_names = [p.name for p in previous]
+
+    src = [get_dlpy_layer(layers, i) for i in src_names]
+
+    return Concat(name=node.name,
+                  act='identity',
+                  src_layers=src)
+
+
+def onnx_extract_gemm(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+        neurons = a_shape[0]
+    
+    return Dense(n=neurons,
+                 name=node.name,
+                 act=act,
+                 include_bias=include_bias,
+                 src_layers=src)
+
+
+def onnx_extract_matmul(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc string

File: dlpy/model_conversion/sas_onnx_parse.py [hunk 13]
```diff
@@ -0,0 +1,855 @@
...
+            include_bias = True
+
+    return Dense(n=neurons,
+                 name=node.name,
+                 act=act,
+                 include_bias=include_bias,
+                 src_layers=src)
+
+
+def onnx_extract_residual(graph, node, layers):
```
Comment (ethem-kinginthenorth) [inline dlpy/model_conversion/sas_onnx_parse.py]: parameters doc striong

**Comment by ethem-kinginthenorth** [comment]
great job @yg2288

**Pull request merged**